src/ol/colorlike.js

```diff
@@ -16,13 +31,70 @@ import {toString} from './color.js';
  */
 
 /**
- * @param {import("./color.js").Color|ColorLike} color Color.
- * @return {ColorLike} The color as an {@link ol/colorlike~ColorLike}.
+ * @param {import("./color.js").Color|ColorLike|PatternDescriptor|null} color Color.
+ * @return {ColorLike|null} The color as an {@link ol/colorlike~ColorLike}.
  * @api
  */
 export function asColorLike(color) {
+  if (!color) {
+    return null;
+  }
+  if (typeof color === 'object' && 'src' in color) {
+    return asCanvasPattern(color);
+  }
   if (Array.isArray(color)) {
     return toString(color);
   }
   return color;
 }
+
+/**
+ * @param {PatternDescriptor} pattern Pattern descriptor.
+ * @return {CanvasPattern|null} Canvas pattern or null if the pattern referenced in the
+ * PatternDescriptor was not found in the icon image cache.
+ */
+function asCanvasPattern(pattern) {
+  if (pattern.offset && pattern.size) {
+    const cacheKey = pattern.src + ':' + pattern.offset;
+    let canvasPattern = iconCache.getPattern(
+      cacheKey,
+      undefined,
+      pattern.color
+    );
```

Comment:
I'm not yet understanding this logic.  If `iconCache.getPattern(cacheKey, undefined, pattern.color)` is truthy, then we return `iconCache.getPattern(pattern.src, 'anonymous', pattern.color)`.

Reply:
Ah, I think I misread the diff.

Inverting some of the conditions and returning early when possible
```js
/**
 * @param {PatternDescriptor} pattern Pattern descriptor.
 * @return {CanvasPattern|null} Canvas pattern or null if the pattern referenced in the
 * PatternDescriptor was not found in the icon image cache.
 */
function asCanvasPattern(pattern) {
  if (!pattern.offset || !pattern.size) {
    return iconCache.getPattern(pattern.src, 'anonymous', pattern.color);
  }

  const cacheKey = pattern.src + ':' + pattern.offset;
  const canvasPattern = iconCache.getPattern(
    cacheKey,
    undefined,
    pattern.color
  );

  if (canvasPattern) {
    return canvasPattern;
  }

  const iconImage = iconCache.get(pattern.src, 'anonymous', null);
  if (iconImage.getImageState() !== ImageState.LOADED) {
    return null;
  }

  const patternCanvasContext = createCanvasContext2D(
    pattern.size[0],
    pattern.size[1]
  );
  patternCanvasContext.drawImage(
    iconImage.getImage(1),
    pattern.offset[0],
    pattern.offset[1],
    pattern.size[0],
    pattern.size[1],
    0,
    0,
    pattern.size[0],
    pattern.size[1]
  );
  getIconImage(
    patternCanvasContext.canvas,
    cacheKey,
    undefined,
    ImageState.LOADED,
    pattern.color,
    true
  );

  return iconCache.getPattern(cacheKey, undefined, pattern.color);
}
```

I think this is the same logic - just easier (for me) to make sense of things with shorter `if` blocks.

Reply:
The former is a cached colored sub-image from a sprite sheet (`pattern.offset` and `pattern.size` are specified - see `if (...` in line 57 with `return ...` in line 97). The latter is a cached colored image directly from the URL specified in `pattern.src` (`return ...` in line 99).